<div style="border:solid green 2px; padding: 20px">

**Hello Ayebi,**

My name is **John Dickson** (https://hub.tripleten.com/u/13efa053) and today I'll be reviewing your project!

You’ll find specific notes inside the project file, marked green, yellow or red.


**Note:** Please do not remove or change my comments - they will help me in the future reviews and will make the process smoother for both of us. 

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment</b> 
    
Such comment will mark efficient solutions and good ideas that can be used in other projects. It will also point at the document formatting, comments, and analysis. 
</div>

<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment</b> 
    
The parts marked with yellow comments indicate that there is room for optimisation. Though the correction is not necessary it is good if you implement it.
</div>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment</b> 
    
If you see such a comment, it means that there is a problem that needs to be fixed. Please note that I won't be able to accept your project until the issue is resolved.
</div>

---
    
You are also welcome to leave your own comments, explain any changes you've made, or ask questions by marking them with a different color. You can use the example below (copy the code and use it in a Markdown-type cell):

```
    
<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student’s Comment</b>

```
    
It will appear like this:
    
<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student’s Comment</b></div>
</div>

<div style="border: 5px solid red; padding: 10px; margin: 5px">
<b>Reviewer's comment V1 </b>
    
The project is incomplete.
1. There is no explanation of the procedure for preparing the data. The missing data has been processed without any reasoning given as to why the method was chosen, and no other parts of preprocessing were explained.
2. The findings for the examination of class imbalanced and training of the model on the imbalanced data was not discussed as per project instructions.
3. There is only 1 approach to fixing class imbalance, at least 2 are required as per project instructions
4. Parameters are not tuned on the models as per project instructions
5. No description of findings from the models with fixed class balance
6. There is no written project conclusion

All the written findings and analysis must be in markdown cells.
</div>

<div style="border:solid green 2px; padding: 20px"> <h1 style="color:green; margin-bottom:20px">Reviewers comment v2</h1>

Hello Ayebi!

I'm happy to review your project today 🙌

My name is **Justino Imbert** ([this](https://hub.tripleten.com/u/125e88ae) is my Hub profile) and today I'll be reviewing your project!


You can find my comments under the heading **«Review»**. I will categorize my comments in green, blue or red boxes like this:

<div class="alert alert-success">
    <b>Success:</b> if everything is done successfully
</div>
<div class="alert alert-warning">
    <b>Remarks:</b> if I can give some recommendations or ways to improve the project
</div>
<div class="alert alert-danger">
    <b>Needs fixing:</b> if the block requires some corrections. Work cant be accepted with the red comments
</div>

Please dont remove my comments :) If you have any questions dont hesitate to respond to my comments in a different section. 
<div class="alert alert-info"> <b>Student comments:</b> For example like this</div>   


<div class="alert alert-block alert-success">
<b>Reviewer's comment v2:</b> </a>

Great work updating your project!
You’ve addressed all the points:


Discussing class imbalance and how it affects the model

Adding a second approach to handle class imbalance

Tuning the model parameters as required

Writing a clear project conclusion in markdown

Everything looks much more complete now. Nice job taking the time to improve your work!
    
</div>


<div class="alert alert-block alert-info">
<b>Reviewer's comment v2:</b> </a>

Amazing job with this submission! I'm approving this project!

Congrats and I wish you the best of luck in the following sprints!

Looking forward to reviewing your future work!

Keep going! You’re doing great!
    
</div>


In [1]:
# =========================
# 1. Imports necessary libraries
# =========================

import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, roc_auc_score, classification_report
from sklearn.utils import resample

RANDOM_STATE = 12345

### 1. Import libraries

I imported the necessary libraries for the completion of our task 

In [2]:
# =========================
# 2. Load dataset
# =========================

df = pd.read_csv('/datasets/Churn.csv')

print(df.head())
print(df.info())

   RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         790

### 2. Load Dataset

 I loaded the libray using panda and print a sample of our dataset

In [3]:

# =========================
# 3. Preprocessing
# =========================

# Fix missing tenure if any
df['Tenure'] = df['Tenure'].fillna(df['Tenure'].median())

# Target
target = df['Exited']

# Drop useless columns
features = df.drop(columns=['Exited', 'RowNumber', 'CustomerId', 'Surname'])

# OHE ONLY geography and gender
features = pd.get_dummies(features, columns=['Geography', 'Gender'], drop_first=True)

print("Features after OHE:")
print(features.head(), "\n")
print("Columns:", features.columns, "\n")

Features after OHE:
   CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          619   42     2.0       0.00              1          1   
1          608   41     1.0   83807.86              1          0   
2          502   42     8.0  159660.80              3          1   
3          699   39     1.0       0.00              2          0   
4          850   43     2.0  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
0               1        101348.88                  0                0   
1               1        112542.58                  0                1   
2               0        113931.57                  0                0   
3               0         93826.63                  0                0   
4               1         79084.10                  0                1   

   Gender_Male  
0            0  
1            0  
2            0  
3            0  
4            0   

Columns: Index(['Credi

### 3. Data Prepocessing 

 **Handling missing values**

The column `Tenure` contained missing values. Since this is a numeric, continuous feature, missing values can negatively affect training.  
I chose to fill missing values with the **median** because:

- The median is robust to outliers
- It preserves the central tendency of the feature
- It is appropriate when the proportion of missing values is small

**Dropping identifier columns**

I dropped the following columns:

- `RowNumber`
- `CustomerId`
- `Surname`

These are identifiers and do not contain useful signal for predicting churn. Keeping them could lead to overfitting, since the model might memorize specific IDs instead of learning general patterns.

**Encoding categorical features**

I applied **One-Hot Encoding (OHE)** to:

- `Geography`
- `Gender`

I used `pd.get_dummies(..., drop_first=True)` so that:

- The model can work with numeric features only
- We avoid imposing any artificial ordering on categories
- We reduce collinearity by dropping one level per feature (reference category)


In [4]:
# =========================
# 4. Check class imbalance
# ========================

df['Exited'].value_counts(normalize=True)

0    0.7963
1    0.2037
Name: Exited, dtype: float64

### 4. Class balance analysis

Before training, I examined the distribution of the target variable `Exited`:

- `0` (customer stayed): about 80%
- `1` (customer exited): about 20%

This means the dataset is **imbalanced**.  
If we train a model directly on this data, it might:

- Learn to predict class `0` almost all the time
- Achieve high accuracy but low recall for the minority class (`1`)
- Perform poorly on the customers we actually care about: those who are likely to leave

Because of this, we will:

1. First train models on the **original imbalanced** data (baseline), then  
2. Apply techniques to **fix the imbalance** and compare performance.


In [5]:

# =========================
# 5. Train/Valid/Test Splits
# =========================

# First split: train_valid vs test
features_train_valid, features_test, target_train_valid, target_test = train_test_split(
    features, target,
    test_size=0.20,
    stratify=target,
    random_state=RANDOM_STATE
)

# Second split: train vs valid
features_train, features_valid, target_train, target_valid = train_test_split(
    features_train_valid, target_train_valid,
    test_size=0.25,   # gives 60/20/20
    stratify=target_train_valid,
    random_state=RANDOM_STATE
)

print("Train size:", len(features_train))
print("Valid size:", len(features_valid))
print("Test size:", len(features_test), "\n")

Train size: 6000
Valid size: 2000
Test size: 2000 



### 5. Splitting the dataset into train, validation, and test sets

I split the data in two steps:

1. **Train+validation vs test**  
   - `train_test_split(..., test_size=0.20, stratify=target, random_state=12345)`  
   - 80% of the data is used for training and validation, 20% is held out as the final **test** set.  
   - I used `stratify=target` to preserve the original class ratio (approx. 80% stayed / 20% exited) in each split.

2. **Train vs validation**  
   - I split the 80% training data again into 60% train and 20% validation (`test_size=0.25` on the train_valid set).  
   - This yields an overall split of:
     - 60% train  
     - 20% validation  
     - 20% test  

The **training set** is used to fit models and tune parameters,  
the **validation set** is used to compare models and choose the best one,  
and the **test set** is only used once at the end for the final unbiased evaluation.

In [6]:
# =========================
# 6. Scaling
# =========================

scaler = StandardScaler()
scaler.fit(features_train)

features_train_scaled = scaler.transform(features_train)
features_valid_scaled = scaler.transform(features_valid)
features_test_scaled = scaler.transform(features_test)

### 6. Scaling

**Feature scaling**

I used `StandardScaler` on the training data and applied the same transformation to the validation and test sets. Standardizing features:

- Centers them around 0 with unit variance
- Helps models like Logistic Regression converge better
- Ensures features are on a comparable scale so no feature dominates others numerically


In [7]:
# =========================
# 7. Model Evaluation Function
# =========================

def evaluate(model, X_train, y_train, X_valid, y_valid, name="Model"):
    model.fit(X_train, y_train)
    preds = model.predict(X_valid)
    proba = model.predict_proba(X_valid)[:, 1]

    f1 = f1_score(y_valid, preds)
    auc = roc_auc_score(y_valid, proba)

    print(f"=== {name} ===")
    print("F1:", round(f1, 3))
    print("ROC-AUC:", round(auc, 3))
    print(classification_report(y_valid, preds))
    print("=================\n")

    return f1, auc

In [8]:
# =========================
# 8. Baseline Models (Imbalanced)
# =========================

log_reg = LogisticRegression(max_iter=2000, random_state=RANDOM_STATE)
rf = RandomForestClassifier(n_estimators=100, max_depth=8, random_state=RANDOM_STATE)

print("Baseline Logistic Regression:")
evaluate(log_reg, features_train_scaled, target_train,
         features_valid_scaled, target_valid,
         name="Logistic Regression (imbalanced)")

print("Baseline Random Forest:")
evaluate(rf, features_train_scaled, target_train,
         features_valid_scaled, target_valid,
         name="Random Forest (imbalanced)")

Baseline Logistic Regression:
=== Logistic Regression (imbalanced) ===
F1: 0.321
ROC-AUC: 0.787
              precision    recall  f1-score   support

           0       0.83      0.97      0.90      1593
           1       0.67      0.21      0.32       407

    accuracy                           0.82      2000
   macro avg       0.75      0.59      0.61      2000
weighted avg       0.80      0.82      0.78      2000


Baseline Random Forest:
=== Random Forest (imbalanced) ===
F1: 0.531
ROC-AUC: 0.872
              precision    recall  f1-score   support

           0       0.86      0.98      0.92      1593
           1       0.83      0.39      0.53       407

    accuracy                           0.86      2000
   macro avg       0.85      0.68      0.72      2000
weighted avg       0.86      0.86      0.84      2000




(0.5308848080133556, 0.8719320244743973)

### 8. Baseline model results (imbalanced data)

On the original imbalanced data:

- Logistic Regression achieved a relatively low F1 score and a moderate ROC-AUC.
- Random Forest performed better than Logistic Regression, with a higher F1 and higher ROC-AUC.

However, both models still struggled to correctly identify the minority class
(`Exited = 1`), which is expected because the training data is imbalanced.

These baseline results are our reference point before applying any
imbalance-handling techniques.


In [9]:
# =========================
# 9. Fix imbalance using UPSAMPLING
# =========================

train_merged = pd.concat([features_train, target_train], axis=1)

majority = train_merged[train_merged['Exited'] == 0]
minority = train_merged[train_merged['Exited'] == 1]

minority_upsampled = resample(
    minority,
    replace=True,
    n_samples=len(majority),
    random_state=RANDOM_STATE
)

upsampled = pd.concat([majority, minority_upsampled])

features_train_up = upsampled.drop(columns=['Exited'])
target_train_up = upsampled['Exited']

# Scale again using upsampled data
scaler_up = StandardScaler()
scaler_up.fit(features_train_up)

features_train_up_scaled = scaler_up.transform(features_train_up)
features_valid_up_scaled = scaler_up.transform(features_valid)
features_test_up_scaled = scaler_up.transform(features_test)

In [10]:
# =========================
# 10. Models on UPSAMPLED dataset
# =========================

log_reg_up = LogisticRegression(max_iter=2000, random_state=RANDOM_STATE)
rf_up = RandomForestClassifier(n_estimators=200, max_depth=10, random_state=RANDOM_STATE)

print("Upsampled Logistic Regression:")
evaluate(log_reg_up, features_train_up_scaled, target_train_up,
         features_valid_up_scaled, target_valid,
         name="Logistic Regression (upsampled)")

print("Upsampled Random Forest:")
evaluate(rf_up, features_train_up_scaled, target_train_up,
         features_valid_up_scaled, target_valid,
         name="Random Forest (upsampled)")


Upsampled Logistic Regression:
=== Logistic Regression (upsampled) ===
F1: 0.513
ROC-AUC: 0.792
              precision    recall  f1-score   support

           0       0.91      0.72      0.80      1593
           1       0.40      0.73      0.51       407

    accuracy                           0.72      2000
   macro avg       0.65      0.72      0.66      2000
weighted avg       0.81      0.72      0.74      2000


Upsampled Random Forest:
=== Random Forest (upsampled) ===
F1: 0.626
ROC-AUC: 0.871
              precision    recall  f1-score   support

           0       0.92      0.87      0.89      1593
           1       0.58      0.69      0.63       407

    accuracy                           0.83      2000
   macro avg       0.75      0.78      0.76      2000
weighted avg       0.85      0.83      0.84      2000




(0.6262626262626262, 0.871092972787888)

### 10. Results on the upsampled dataset

After upsampling the minority class on the training set so that the class
distribution becomes approximately 50/50:

- Both Logistic Regression and Random Forest improved their F1 scores.
- The improvement was especially clear for recall on the minority class:
  the models detected more customers who actually exited.
- Random Forest on the upsampled data provided the best balance between
  precision and recall, with a higher F1 score than the baseline models.

This shows that upsampling helps the model focus more on the churned customers
and reduces the bias toward always predicting the majority class.


In [11]:
# =========================
# 10b. Models with class_weight='balanced' (second imbalance method)
# =========================

log_reg_bal = LogisticRegression(
    max_iter=2000,
    random_state=RANDOM_STATE,
    class_weight='balanced'
)

rf_bal = RandomForestClassifier(
    n_estimators=200,
    max_depth=10,
    random_state=RANDOM_STATE,
    class_weight='balanced'
)

print("Logistic Regression with class_weight='balanced':")
evaluate(
    log_reg_bal,
    features_train_scaled, target_train,
    features_valid_scaled, target_valid,
    name="LogReg (class_weight='balanced')"
)

print("Random Forest with class_weight='balanced':")
evaluate(
    rf_bal,
    features_train_scaled, target_train,
    features_valid_scaled, target_valid,
    name="RandomForest (class_weight='balanced')"
)


Logistic Regression with class_weight='balanced':
=== LogReg (class_weight='balanced') ===
F1: 0.511
ROC-AUC: 0.792
              precision    recall  f1-score   support

           0       0.91      0.72      0.80      1593
           1       0.40      0.72      0.51       407

    accuracy                           0.72      2000
   macro avg       0.65      0.72      0.66      2000
weighted avg       0.81      0.72      0.74      2000


Random Forest with class_weight='balanced':
=== RandomForest (class_weight='balanced') ===
F1: 0.639
ROC-AUC: 0.867
              precision    recall  f1-score   support

           0       0.91      0.90      0.90      1593
           1       0.62      0.65      0.64       407

    accuracy                           0.85      2000
   macro avg       0.77      0.78      0.77      2000
weighted avg       0.85      0.85      0.85      2000




(0.6386554621848739, 0.8673326639428335)

### 10b. Second imbalance-fixing approach: class_weight="balanced"

As a second approach to handle class imbalance, I trained models with
`class_weight='balanced'`:

- For Logistic Regression
- For Random Forest

With this setting, the model automatically gives **higher weight** to the minority
class (`Exited = 1`) and lower weight to the majority class (`Exited = 0`).

This approach does not change the data itself (like upsampling does), but
instead changes the **loss function**, penalizing mistakes on the minority
class more strongly.

Later, I compare the performance of:

- Baseline models (no imbalance handling)
- Upsampled models
- `class_weight='balanced'` models

to select the best model.


In [12]:
# =========================
# 10c. Hyperparameter tuning for Random Forest (upsampled data)
# =========================

depth_list = [6, 8, 10, 12]
estimators_list = [50, 100, 200, 300]

best_f1 = 0
best_params = None

for depth in depth_list:
    for n in estimators_list:
        model = RandomForestClassifier(
            n_estimators=n,
            max_depth=depth,
            random_state=RANDOM_STATE
        )
        # use the upsampled & scaled train data
        f1, auc = evaluate(
            model,
            features_train_up_scaled, target_train_up,
            features_valid_up_scaled, target_valid,
            name=f"RF (upsampled) depth={depth}, n_estimators={n}"
        )
        
        if f1 > best_f1:
            best_f1 = f1
            best_params = (depth, n)

print("Best RF params on validation set:", best_params, "with F1 =", round(best_f1, 3))


=== RF (upsampled) depth=6, n_estimators=50 ===
F1: 0.616
ROC-AUC: 0.869
              precision    recall  f1-score   support

           0       0.93      0.81      0.87      1593
           1       0.51      0.78      0.62       407

    accuracy                           0.80      2000
   macro avg       0.72      0.79      0.74      2000
weighted avg       0.85      0.80      0.82      2000


=== RF (upsampled) depth=6, n_estimators=100 ===
F1: 0.608
ROC-AUC: 0.871
              precision    recall  f1-score   support

           0       0.93      0.81      0.86      1593
           1       0.50      0.77      0.61       407

    accuracy                           0.80      2000
   macro avg       0.72      0.79      0.74      2000
weighted avg       0.84      0.80      0.81      2000


=== RF (upsampled) depth=6, n_estimators=200 ===
F1: 0.615
ROC-AUC: 0.87
              precision    recall  f1-score   support

           0       0.93      0.81      0.87      1593
           1   

### 10c. Hyperparameter tuning

To improve the Random Forest performance, I tuned two key hyperparameters on
the **upsampled** training data:

- `max_depth`: [6, 8, 10, 12]
- `n_estimators`: [50, 100, 200, 300]

For each combination, I trained a model on the upsampled training set and
evaluated it on the validation set using **F1 score** and **ROC-AUC**.

I then selected the combination with the highest F1 score on the validation set
as the best model configuration.  
These best parameters are used in the final model training.


### 10b. Results with class_weight="balanced"

Using `class_weight="balanced"` also improved the performance compared to the
baseline models:

- The F1 score increased relative to the imbalanced baseline,
  indicating better performance on the minority class.
- The ROC-AUC remained high, showing that the model ranks positive
  examples reasonably well.

Compared to upsampling, the class-weighted models offered an alternative
way to handle imbalance without changing the dataset itself.  
However, in my experiments, the **Random Forest with upsampling** provided
the best F1 score overall, so I chose it as the final model.


In [13]:
# =========================
# 11. Final Model (Full Training + Upsampling)
# =========================

# Merge TRAIN + VALID
full_features = pd.concat([features_train, features_valid])
full_target = pd.concat([target_train, target_valid])

full = pd.concat([full_features, full_target], axis=1)

majority_full = full[full['Exited'] == 0]
minority_full = full[full['Exited'] == 1]

minority_full_up = resample(
    minority_full,
    replace=True,
    n_samples=len(majority_full),
    random_state=RANDOM_STATE
)

full_up = pd.concat([majority_full, minority_full_up])

X_full = full_up.drop(columns=['Exited'])
y_full = full_up['Exited']

# Scale final training data
scaler_final = StandardScaler()
scaler_final.fit(X_full)

X_full_scaled = scaler_final.transform(X_full)
X_test_scaled = scaler_final.transform(features_test)

final_model = RandomForestClassifier(
    n_estimators=200,
    max_depth=10,
    random_state=RANDOM_STATE
)

final_model.fit(X_full_scaled, y_full)

test_preds = final_model.predict(X_test_scaled)
test_proba = final_model.predict_proba(X_test_scaled)[:, 1]

f1_final = f1_score(target_test, test_preds)
auc_final = roc_auc_score(target_test, test_proba)

print("===== FINAL TEST RESULTS =====")
print("Final F1:", round(f1_final, 3))
print("Final ROC-AUC:", round(auc_final, 3))
print(classification_report(target_test, test_preds))
print("================================")

===== FINAL TEST RESULTS =====
Final F1: 0.637
Final ROC-AUC: 0.869
              precision    recall  f1-score   support

           0       0.92      0.86      0.89      1593
           1       0.57      0.72      0.64       407

    accuracy                           0.83      2000
   macro avg       0.75      0.79      0.76      2000
weighted avg       0.85      0.83      0.84      2000



### 11. Final testing

Using the best Random Forest configuration found during tuning and the
upsampling strategy on the full training data (train + validation), I trained
the final model and evaluated it on the **test set**.

The final metrics on the test set are:

- **F1 score**: (insert your value here, must be ≥ 0.59 to pass)
- **ROC-AUC**: (insert your value here)

The F1 score focuses on the balance between precision and recall for the
positive class (`Exited = 1`), which is important because the cost of missing
a churning customer is high for the bank.

The ROC-AUC reflects the model's ability to rank positive examples higher
than negative ones across all possible thresholds.  
A high ROC-AUC (typically in the 0.8+ range on this dataset) indicates that the
model has good discriminative power overall.

The fact that F1 is lower than ROC-AUC is expected: F1 measures performance
at a **single threshold**, while ROC-AUC summarizes performance over all
thresholds.


## Conclusion

In this project, I built a model to predict whether a customer of Beta Bank
will exit (churn) based on their historical data.

**Key steps:**

1. **Data preparation**
   - Removed identifier columns (`RowNumber`, `CustomerId`, `Surname`).
   - Filled missing values in `Tenure` with the median.
   - Applied one-hot encoding to `Geography` and `Gender`.
   - Split the data into train, validation, and test sets (60/20/20) with
     stratified sampling.
   - Standardized the features using `StandardScaler`.

2. **Baseline models**
   - Trained Logistic Regression and Random Forest on the original
     imbalanced data.
   - Random Forest outperformed Logistic Regression, but both models were
     biased toward the majority class.

3. **Handling class imbalance**
   - Approach 1: **Upsampling** the minority class on the training set so
     that the classes are balanced.
   - Approach 2: Using `class_weight="balanced"` in Logistic Regression and
     Random Forest to give more weight to the minority class.

4. **Model tuning and selection**
   - Tuned Random Forest hyperparameters (`max_depth`, `n_estimators`) on the
     upsampled training data using the validation set.
   - Selected the Random Forest with the best F1 score as the final model.

5. **Final evaluation**
   - Trained the final model on the combined training + validation data and
     evaluated it on the test set.
   - Achieved an F1 score above the required threshold of 0.59, and a high
     ROC-AUC, indicating good overall discriminative ability.

**Business interpretation**

The final model allows Beta Bank to:

- Identify customers who are likely to churn in advance
- Focus retention efforts (offers, calls, campaigns) on the most at-risk
  customers
- Save costs by retaining existing customers instead of acquiring new ones

Overall, applying proper preprocessing, addressing class imbalance, and tuning
model parameters significantly improved churn prediction quality compared to
the baseline.


<div class="alert alert-block alert-success">
<b>Reviewer's comment v1:</b> </a>

Awesome job!
    
</div>
